packages/salesforcedx-vscode-replay-debugger/src/states/variableAssignmentState.ts

```diff
@@ -26,106 +26,150 @@ export class VariableAssignmentState implements DebugLogState {
       } else if (frameInfo.locals.has(varName)) {
-        // if name does not contain '.' (i.e. this.attr or a.Name), it should be in locals and we can update the value
         map = frameInfo.locals;
         container = map.get(varName) as ApexVariableContainer;
+        // if the variable we're given is a child variable, then it will come in the format of this.varName
+      } else if (name.indexOf('.') !== -1) {
+        isNested = true;
       }
 
-      if (
-        container &&
-        map &&
-        container.variablesRef !== 0 &&
-        !logContext.getRefsMap().has(ref)
-      ) {
-        container = new ApexVariableContainer(
-          container.name,
-          '',
-          container.type
-        );
-        map.set(varName, container);
-      }
-
-      if (container) {
-        container.value = value;
-      }
-
-      // assigning to top level variable in locals
-      if (
-        ref !== '0' &&
-        container &&
-        !container.type.startsWith('Map<') &&
-        !container.type.startsWith('List<') &&
-        !container.type.startsWith('Set<')
-      ) {
-        if (!logContext.getRefsMap().has(ref)) {
-          logContext.getRefsMap().set(ref, container);
-        } else {
-          // if the ref already exists then that means we are assigning another pointer to a ref so we need to merge the info from the ref into the variable container
-          const refContainer = logContext.getRefsMap().get(ref)!;
-          container.variables = refContainer.variables;
-          container.variablesRef = refContainer.variablesRef;
-          container.name = varName;
+      // update the ref mapping
+      if (ref) {
+        if (!refMap.has(ref)) {
+          logContext
+            .getRefsMap()
+            .set(ref, new ApexVariableContainer('', '', '', ref));
         }
-        if (value.indexOf('{') === 0 && value !== '{}') {
-          container.value = '';
-          container.variablesRef = logContext
-            .getVariableHandler()
-            .create(container);
-          this.parseAndPopulate(value, container);
+        const refContainer = refMap.get(ref)!;
+        if (value !== '{}') {
+          // nested variable will either be given a json or a value
+          if (isNested) {
+            // if its a json assignment, parse the values
+            if (value.indexOf('{') === 0) {
+              const topLevel = new ApexVariableContainer(varName, '', '');
+              refContainer.variables.set(varName, topLevel);
+              topLevel.variablesRef = logContext
+                .getVariableHandler()
+                .create(topLevel);
+              this.parseJSONAndPopulate(value, topLevel, logContext);
+            } else {
+              // if it's not nested then we check if the value is a reference
+              if (refMap.has(value)) {
+                const pulledRef = refMap.get(value) as ApexVariableContainer;
+                const tmpContainer = new ApexVariableContainer(
```

Comment:
I see you copy containers here and in `parseJSONAndPopulate`. Could you pull that out to a common method, maybe in ApexVariableContainer? If the parts that are copied change in the future, we can be rest assured both are changed.